scripts/ts/menu_store.ts

```diff
@@ -0,0 +1,217 @@
+     */
+    setActive(element: HTMLElement): void {
+      if (this.store.indexOf(element) !== -1) {
+        this.active = element;
+      }
+    }
+
+    /**
+     * @return {HTMLElement} The currently active store element, if one exists.
+     */
+    getActive(): HTMLElement {
+      return this.active;
+    }
+
+    /**
+     * Returns next active element.
+     * If store is empty returns null and also unsets active element.
+     * If active is not set returns the first element of the store.
+     * @return {HTMLElement} The next element if it exists.
+     */
+    next(): HTMLElement {
+      let length = this.store.length;
+      if (length === 0) {
+        this.active = null;
+        return null;
+      }
+      let index = this.store.indexOf(this.active);
+      index = index === -1 ? 0 : (index < length - 1 ? index + 1 : 0);
+      this.active = this.store[index];
+      return this.active;
+    }
+
+    /**
+     * Returns next active element.
+     * If store is empty returns null and also unsets active element.
+     * If active is not set returns the first element of the store.
+     * @return {HTMLElement} The next element if it exists.
+     */
+    previous(): HTMLElement {
+      let length = this.store.length;
+      if (length === 0) {
+        this.active = null;
+        return null;
+      }
+      let last = length - 1;
+      let index = this.store.indexOf(this.active);
+      index = index === -1 ? last : (index === 0 ? last : index - 1);
+      this.active = this.store[index];
+      return this.active;
+    }
+
+    //// TODO: Implement with Promises?
+    replace(oldElement: HTMLElement, newElement: HTMLElement): void {
+      let index = this.store.indexOf(oldElement);
+      index === -1 ? this.store.push(newElement) :
+        this.store.splice(index, 1, newElement);
+    }
+
+    insert(element: HTMLElement) {
+      this.insertAt(element, this.store.length);
+    }
+
+    insertAll(elements: HTMLElement[]) {
+      this.removeAll();
+      for (let i = elements.length - 1, element: HTMLElement;
+           element = elements[i]; i--) {
+        this.insertAt(element, 0);
+      }
+    }
+
+    insertBefore(oldElement: HTMLElement, newElement: HTMLElement) {
+      let index = this.store.indexOf(oldElement);
+      index === -1 ? this.insertAt(newElement, 0) :
+        this.insertAt(newElement, index);
+      }
+
+    insertAt(element: HTMLElement, position: number) {
+      // Twice?
+      this.addEvents(element);
+      this.store.splice(position, 0, element);
+    }
+
+    remove(element: HTMLElement) {
+      this.removeAt(this.store.indexOf(element));
+    }
+
+    removeAll() {
+      while (this.store.length > 0) {
+        this.removeAt(0);
+      }
+    }
+
+    //// TODO: We somehow should put those in by taborder.  Maybe add a special
+    // class name and do a document.  Would make it also easier to see if an
+    // element has already events attached.
+    //
```

Comment:
We also need to take care of uniqueness per context menu to ensure multiple context menus can be used. Have some global self-incrementing counter.